cli_tools/diagnostics/main_windows.go

```diff
@@ -0,0 +1,241 @@
+//  Copyright 2018 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	sys32 = `C:\Windows\System32`
+)
+
+type runner interface {
+	run() (string, error)
+}
+
+type cmd struct {
+	name           string
+	args           string
+	outputFileName string
+	// True when the command produces its own file and doesn't need one
+	// created from stdout.
+	cmdProducesFile bool
+}
+
+type wmiQuery struct {
+	class          string
+	namespace      string
+	outputFileName string
+}
+
+func (command cmd) run() (string, error) {
+	outPath := filepath.Join(tmpFolder, command.outputFileName)
+
+	cmdPath := filepath.Join(sys32, command.name)
+	c := exec.Command(cmdPath)
+	argString := command.args
+
+	if command.cmdProducesFile {
+		// Replace any output file args with that path in a temp folder
+		relPath := command.outputFileName
+		argString = strings.Replace(argString, relPath, outPath, -1)
+	} else {
+		// If the command doesn't produce a file, we need to construct
+		// one from Stdout and Stderr
+		outFile, err := os.Create(outPath)
+		if err != nil {
+			log.Printf("Error creating file %s: %v", outPath, err)
+			return outPath, err
+		}
+		defer outFile.Close()
+		c.Stdout = outFile
+		c.Stderr = outFile
+	}
+
+	if command.args != "" {
+		c.Args = append(c.Args, strings.Split(argString, " ")...)
+	}
+	err := c.Run()
+	return outPath, err
+}
+
+func (query wmiQuery) run() (string, error) {
+	outPath := filepath.Join(tmpFolder, query.outputFileName)
+	outFile, err := os.Create(outPath)
+	if err != nil {
+		return outPath, err
+	}
+	defer outFile.Close()
+
+	// WMI is somewhat flaky, so we should retry a few times on failures
+	var data string
+	for i := 0; i < 3; i++ {
+		data, err = printWmiObjects(query.class, query.namespace)
+		if err == nil {
+			break
+		}
+	}
+	if err != nil {
+		return outPath, err
+	}
+
+	header := fmt.Sprintf("Queried wmi objects [%s] from namespace %s\n\n", query.class, query.namespace)
+	_, err = outFile.WriteString(header)
```

Comment:
if _, err := ...; err != nil {

Reply:
Fixed this everywhere I've seen it. 

Just for clarity, you only do this if the value(s) being returned are being used exclusively within the scope of the if statement?

Reply:
Yes, the if statement scopes the variable declaration